// sanity-plugin-iframe-pane/lib/index.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: !0 });
var jsxRuntime = require("react/jsx-runtime"), icons = require("@sanity/icons"), createSecret = require("@sanity/preview-url-secret/create-secret"), definePreviewUrl = require("@sanity/preview-url-secret/define-preview-url"), ui = require("@sanity/ui"), framerMotion = require("framer-motion"), react = require("react"), sanity = require("sanity"), suspendReact = require("suspend-react"), getRedirectTo = require("@sanity/preview-url-secret/get-redirect-to");
function DisplayUrl(props) {
  const truncatedUrl = react.useMemo(() => {
    const url = getRedirectTo.getRedirectTo(props.url);
    return `${url.origin === location.origin ? "" : url.origin}${url.pathname}${url.search}`;
  }, [props.url]);
  return /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: 0, textOverflow: "ellipsis", children: truncatedUrl });
}
const sizes = {
  desktop: {
    width: "100%",
    height: "100%"
  },
  mobile: {
    width: 414,
    height: 746
  }
}, DEFAULT_SIZE = "desktop";
function Toolbar(props) {
  const { url, iframeSize, setIframeSize, reloading, showUrl, reloadButton, handleReload } = props, validUrl = url instanceof URL, input = react.useRef(null), { push: pushToast } = ui.useToast(), [, copy] = useCopyToClipboard();
  return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      "textarea",
      {
        style: { position: "absolute", pointerEvents: "none", opacity: 0 },
        ref: input,
        value: validUrl ? url.toString() : "",
        readOnly: !0,
        tabIndex: -1
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { padding: 2, borderBottom: !0, children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { align: "center", gap: 2, children: [
      /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { align: "center", gap: 1, children: /* @__PURE__ */ jsxRuntime.jsx(
        ui.Tooltip,
        {
          animate: !0,
          content: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: 1, style: { whiteSpace: "nowrap" }, children: iframeSize === "mobile" ? "Exit mobile preview" : "Preview mobile viewport" }),
          padding: 2,
          placement: "bottom-start",
          children: /* @__PURE__ */ jsxRuntime.jsx(
            ui.Button,
            {
              disabled: !validUrl,
              fontSize: [1],
              padding: 2,
              mode: iframeSize === "mobile" ? "default" : "ghost",
              icon: icons.MobileDeviceIcon,
              onClick: () => setIframeSize(iframeSize === "mobile" ? "desktop" : "mobile")
            }
          )
        }
      ) }),
      /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { flex: 1, children: showUrl && validUrl && /* @__PURE__ */ jsxRuntime.jsx(DisplayUrl, { url }) }),
      /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { align: "center", gap: 1, children: [
        reloadButton ? /* @__PURE__ */ jsxRuntime.jsx(
          ui.Tooltip,
          {
            animate: !0,
            content: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: 1, style: { whiteSpace: "nowrap" }, children: reloading ? "Reloading\u2026" : "Reload" }),
            padding: 2,
            children: /* @__PURE__ */ jsxRuntime.jsx(
              ui.Button,
              {
                disabled: !validUrl,
                mode: "bleed",
                fontSize: [1],
                padding: 2,
                icon: icons.RefreshIcon,
                loading: reloading,
                "aria-label": "Reload",
                onClick: () => handleReload()
              }
            )
          }
        ) : null,
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Tooltip,
          {
            animate: !0,
            content: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: 1, style: { whiteSpace: "nowrap" }, children: "Copy URL" }),
            padding: 2,
            children: /* @__PURE__ */ jsxRuntime.jsx(
              ui.Button,
              {
                mode: "bleed",
                disabled: !validUrl,
                fontSize: [1],
                icon: icons.CopyIcon,
                padding: [2],
                "aria-label": "Copy URL",
                onClick: () => {
                  input?.current?.value && copy(input.current.value).then((copied) => {
                    pushToast(copied ? {
                      closable: !0,
                      status: "success",
                      title: "The URL is copied to the clipboard"
                    } : {
                      closable: !0,
                      status: "error",
                      title: "Failed to copy the URL to the clipboard"
                    });
                  });
                }
              }
            )
          }
        ),
        /* @__PURE__ */ jsxRuntime.jsx(
          ui.Tooltip,
          {
            animate: !0,
            content: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: 1, style: { whiteSpace: "nowrap" }, children: "Open URL in a new tab" }),
            padding: 2,
            placement: "bottom-end",
            children: /* @__PURE__ */ jsxRuntime.jsx(
              ui.Button,
              {
                disabled: !validUrl,
                fontSize: [1],
                icon: icons.LaunchIcon,
                mode: "ghost",
                paddingY: [2],
                text: "Open",
                "aria-label": "Open URL in a new tab",
                onClick: validUrl ? () => window.open(url.toString()) : void 0
              }
            )
          }
        )
      ] })
    ] }) })
  ] });
}
function useCopyToClipboard() {
  const [copiedText, setCopiedText] = react.useState(null), copy = react.useCallback(async (text) => {
    if (!navigator?.clipboard)
      return console.warn("Clipboard not supported"), !1;
    try {
      return await navigator.clipboard.writeText(text), setCopiedText(text), !0;
    } catch (error) {
      return console.warn("Copy failed", error), setCopiedText(null), !1;
    }
  }, []);
  return [copiedText, copy];
}
const MotionFlex = framerMotion.motion.create(ui.Flex);
function encodeStudioPerspective(studioPerspective) {
  return Array.isArray(studioPerspective) ? studioPerspective.join(",") : studioPerspective;
}
function Iframe(props) {
  const { document, options } = props, draft = document.draft || document.published || document.displayed, { defaultSize = DEFAULT_SIZE, reload, attributes, showDisplayUrl = !0, key } = options, basePath = sanity.useActiveWorkspace()?.activeWorkspace?.basePath || "/", urlRef = react.useRef(options.url), [draftSnapshot, setDraftSnapshot] = react.useState(() => ({ key, draft }));
  react.useEffect(() => {
    urlRef.current = options.url;
  }, [options.url]), react.useEffect(() => {
    JSON.stringify({ key, draft }) !== JSON.stringify(draftSnapshot) && startTransition(() => setDraftSnapshot({ key, draft }));
  }, [draft, draftSnapshot, key]);
  const currentUser = sanity.useCurrentUser(), client = sanity.useClient({ apiVersion: "2023-10-16" }), [expiresAt, setExpiresAt] = react.useState(), previewSecretRef = react.useRef(void 0), [isResolvingUrl, startTransition] = react.useTransition(), { perspectiveStack, selectedPerspectiveName } = sanity.usePerspective(), perspective = react.useMemo(
    () => ({
      perspectiveStack,
      selectedPerspectiveName
    }),
    [perspectiveStack, selectedPerspectiveName]
  ), url = react.useCallback(
    // eslint-disable-next-line @typescript-eslint/no-shadow
    async (draft2) => {
      if (typeof location > "u")
        return;
      const urlProp = urlRef.current;
      if (typeof urlProp == "string")
        return new URL(urlProp, location.origin);
      if (typeof urlProp == "function") {
        const url2 = await urlProp(draft2, perspective);
        return typeof url2 == "string" ? new URL(url2, location.origin) : url2;
      }
      if (typeof urlProp == "object") {
        const preview = typeof urlProp.preview == "function" ? await urlProp.preview(draft2, perspective) : urlProp.preview;
        if (typeof preview != "string")
          return preview;
        if (!previewSecretRef.current) {
          const { secret, expiresAt: expiresAt2 } = await createSecret.createPreviewSecret(
            client,
            "sanity-plugin-iframe-pane",
            location.href,
            currentUser?.id
          );
          previewSecretRef.current = secret, startTransition(() => setExpiresAt(expiresAt2.getTime()));
        }
        const url2 = await definePreviewUrl.definePreviewUrl({
          origin: urlProp.origin === "same-origin" ? location.origin : urlProp.origin,
          preview,
          draftMode: {
            enable: urlProp.draftMode
          }
        })({
          client,
          previewUrlSecret: previewSecretRef.current,
          previewSearchParam: null,
          studioBasePath: basePath,
          studioPreviewPerspective: encodeStudioPerspective(perspective.perspectiveStack)
        });
        return new URL(url2, location.origin);
      }
    },
    [basePath, client, currentUser?.id, perspective]
  );
  return react.useEffect(() => {
    if (expiresAt) {
      const timeout = setTimeout(
        () => {
          startTransition(() => setExpiresAt(void 0)), previewSecretRef.current = void 0;
        },
        Math.max(0, expiresAt - Date.now())
      );
      return () => clearTimeout(timeout);
    }
  }, [expiresAt]), /* @__PURE__ */ jsxRuntime.jsx(react.Suspense, { fallback: /* @__PURE__ */ jsxRuntime.jsx(Loading, { iframeSize: "desktop" }), children: /* @__PURE__ */ jsxRuntime.jsx(
    IframeInner,
    {
      _key: draftSnapshot.key,
      draftSnapshot: draftSnapshot.draft,
      url,
      isResolvingUrl,
      attributes,
      perspective,
      defaultSize,
      reload,
      showDisplayUrl,
      userId: currentUser?.id
    },
    `${draftSnapshot.key}-${selectedPerspectiveName || "draft"}`
  ) });
}
const IframeInner = react.memo(function(props) {
  const {
    isResolvingUrl,
    defaultSize = DEFAULT_SIZE,
    reload,
    attributes = {},
    showDisplayUrl = !0,
    draftSnapshot,
    userId,
    expiresAt,
    perspective: { selectedPerspectiveName, perspectiveStack },
    _key
  } = props, [iframeSize, setIframeSize] = react.useState(sizes?.[defaultSize] ? defaultSize : DEFAULT_SIZE), prefersReducedMotion = ui.usePrefersReducedMotion(), url = suspendReact.suspend(
    () => props.url(draftSnapshot),
    [
      // Cache based on a few specific conditions
      "sanity-plugin-iframe-pane",
      draftSnapshot,
      selectedPerspectiveName,
      perspectiveStack,
      userId,
      expiresAt,
      _key,
      resolveUUID
    ]
  ), [loading, setLoading] = react.useState(!0), [_reloading, setReloading] = react.useState(!1), reloading = _reloading || isResolvingUrl, iframe = react.useRef(null), handleReload = react.useCallback(() => {
    iframe?.current && (iframe.current.src = iframe.current.src, setReloading(!0));
  }, []);
  return /* @__PURE__ */ jsxRuntime.jsx(framerMotion.MotionConfig, { transition: prefersReducedMotion ? { duration: 0 } : void 0, children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { direction: "column", style: { height: "100%" }, children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      Toolbar,
      {
        url,
        iframeSize,
        reloading,
        setIframeSize,
        showUrl: showDisplayUrl,
        reloadButton: !!reload?.button,
        handleReload
      }
    ),
    url instanceof Error ? /* @__PURE__ */ jsxRuntime.jsx(ErrorCard, { error: url }) : /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { tone: "transparent", style: { height: "100%" }, children: /* @__PURE__ */ jsxRuntime.jsx(
      Frame,
      {
        ref: iframe,
        loading,
        reloading,
        iframeSize,
        setReloading,
        setLoading,
        url,
        attributes
      }
    ) })
  ] }) });
}), Frame = react.forwardRef(function(props, iframe) {
  const { loading, setLoading, iframeSize, attributes, reloading, url, setReloading } = props;
  function handleIframeLoad() {
    setLoading(!1), setReloading(!1), attributes.onLoad && typeof attributes.onLoad == "function" && attributes.onLoad();
  }
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { align: "center", justify: "center", style: { height: "100%", position: "relative" }, children: [
    /* @__PURE__ */ jsxRuntime.jsx(framerMotion.AnimatePresence, { children: !url || loading && /* @__PURE__ */ jsxRuntime.jsx(
      MotionFlex,
      {
        initial: "initial",
        animate: "animate",
        exit: "exit",
        variants: spinnerVariants,
        justify: "center",
        align: "center",
        style: { inset: "0", position: "absolute" },
        children: /* @__PURE__ */ jsxRuntime.jsx(Loading, { iframeSize })
      }
    ) }),
    url && /* @__PURE__ */ jsxRuntime.jsx(
      framerMotion.motion.iframe,
      {
        ref: iframe,
        title: "preview",
        frameBorder: "0",
        style: { maxHeight: "100%" },
        src: url.toString(),
        initial: ["background", iframeSize],
        variants: iframeVariants,
        animate: [
          loading ? "background" : "active",
          reloading ? "reloading" : "idle",
          iframeSize
        ],
        ...attributes,
        onLoad: handleIframeLoad
      }
    )
  ] });
}), spinnerVariants = {
  initial: { opacity: 1 },
  animate: { opacity: [0, 0, 1] },
  exit: { opacity: [1, 0, 0] }
}, iframeVariants = {
  ...sizes,
  desktop: {
    ...sizes.desktop,
    boxShadow: "0 0 0 0px var(--card-shadow-outline-color)"
  },
  mobile: {
    ...sizes.mobile,
    boxShadow: "0 0 0 1px var(--card-shadow-outline-color)"
  },
  background: {
    opacity: 0,
    scale: 1
  },
  idle: {
    scale: 1
  },
  reloading: {
    scale: [1, 1, 1, 0.98]
  },
  active: {
    opacity: [0, 0, 1],
    scale: 1
  }
};
function Loading({ iframeSize }) {
  return /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { style: { ...sizes[iframeSize] }, justify: "center", align: "center", direction: "column", gap: 4, children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Spinner, { muted: !0 }),
    /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { muted: !0, size: 1, children: "Loading\u2026" })
  ] });
}
function ErrorCard({ error }) {
  return /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { height: "fill", children: /* @__PURE__ */ jsxRuntime.jsx(ui.Flex, { align: "center", height: "fill", justify: "center", padding: 4, sizing: "border", children: /* @__PURE__ */ jsxRuntime.jsx(ui.Container, { width: 0, children: /* @__PURE__ */ jsxRuntime.jsx(ui.Card, { padding: 4, radius: 2, shadow: 1, tone: "caution", children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Flex, { children: [
    /* @__PURE__ */ jsxRuntime.jsx(ui.Box, { children: /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { size: 1, children: /* @__PURE__ */ jsxRuntime.jsx(icons.WarningOutlineIcon, {}) }) }),
    /* @__PURE__ */ jsxRuntime.jsxs(ui.Stack, { flex: 1, marginLeft: 3, space: 3, children: [
      /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { as: "h1", size: 1, weight: "bold", children: error.name }),
      /* @__PURE__ */ jsxRuntime.jsx(ui.Text, { as: "p", muted: !0, size: 1, children: error.message })
    ] })
  ] }) }) }) }) });
}
const resolveUUID = Symbol();
exports.Iframe = Iframe;
